#1515 - Cache Avro DatumReader and DatumWriter instances to increase performance

wundi opened the pull request:
This is a PR fixing issue #1515 by caching the avro `DatumReader` and `DatumWriter` instances created when using instances of the `AbstractKafkaAvroDeserializer` and `AbstractKafkaAvroDeserializer` classes. Although the issue was only scoped to poor performance when serializing/deserializing `SpecificRecord` instances this PR applies the same caching logic when working with `GenericRecord` since JMH tests show improvements on deserialization there as well.

I have run benchmarks with the `io.confluent.schemaregistry.benchmark.SerdeBenchmark` JMH benchmark, which at the moment only benchmarks generic records. I have benchmarked using a single as well as 40 threads on my machine by running:
```bash
java -jar benchmark/target/benchmarks.jar -bm thrpt -tu s -f 1 -t 1 -wi 10 -i 10 -w 1 -r 1 -p serializationFormat=AVRO
java -jar benchmark/target/benchmarks.jar -bm thrpt -tu s -f 1 -t 40 -wi 10 -i 10 -w 1 -r 1 -p serializationFormat=AVRO
```

The results are: 
```
-------------------------- GenericRecord --------------------------

---- 1 thread ----
-- Stock
Benchmark                   (serializationFormat)   Mode  Cnt        Score        Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  3964313,028 ±  84201,791  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  4878370,388 ± 100343,298  ops/s

-- Caching
Benchmark                   (serializationFormat)   Mode  Cnt        Score        Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  6364262,919 ± 273945,502  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  5127353,620 ± 169815,194  ops/s

---- 40 threads ----
-- Stock
Benchmark                   (serializationFormat)   Mode  Cnt         Score         Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  16769459,760 ±  233665,578  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  31693445,711 ± 1066292,286  ops/s

-- Caching
Benchmark                   (serializationFormat)   Mode  Cnt         Score         Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  33163621,219 ± 1767851,714  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  31070235,686 ±  902265,617  ops/s
```

Single-threaded deserialization is 1.5x faster and multithreaded is 2x faster (on my machine). Serialization is roughly the same.

I have altered the above JMH benchmark to do specific avro serialization and deserialization to assess the converns raised in #1515. That code is not in a commitable state but can be viewed [here](https://github.com/TeletronicsDotAe/schema-registry/blob/issues/1515/master-specific-benchmark/benchmark/src/main/java/io/confluent/schemaregistry/benchmark/SerdeBenchmark.java). Results when running on my machine:

```
-------------------------- SpecificRecord --------------------------

---- 1 thread ----
-- Stock
Benchmark                   (serializationFormat)   Mode  Cnt       Score       Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  358504,030 ± 13906,565  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  379361,650 ± 11641,322  ops/s

-- Caching
Benchmark                   (serializationFormat)   Mode  Cnt        Score       Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  5995665,350 ± 87987,096  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  5107737,779 ± 99491,447  ops/s

---- 40 threads ----
-- Stock
Benchmark                   (serializationFormat)   Mode  Cnt        Score       Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  1781519,126 ± 92044,391  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  1867259,722 ± 52876,002  ops/s

-- Caching
Benchmark                   (serializationFormat)   Mode  Cnt         Score         Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  31863061,274 ±  969895,866  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  29982446,295 ± 1152760,094  ops/s
```

The results are an order of magnitude better, bringing specific avro serialization and deserialization on par with generic.

wundi commented:
[clabot:check]

rayokota commented:
@wundi , actually I think using the record name(s) for the key will not be sufficient, as we need to account for schema evolution.  You will probably need to use a class that contains a pair of Schema instances as the key for the deserializer and a Schema instance as the key for the serializer.

wundi commented:
@rayokota - Did the push but ran out of time before finishing the comment. Appreciate the merge :-) 

And yes, you're right. I implemented the changes you suggested and added a test specifically for schema evolution, that passed after my latest commit but fails on the previous solution. 

There is a test already to assert that projection works, but in that one the full name of the schema changes which isn't the case here. I added the test where I though it had most cohesion with the other tests.

In relation to performance, I'll admit I was a bit worried about the implications of changing caching to use schemas, but the actual numbers look good. Testing in the same fashion I get:

```
-------------------------- GenericRecord --------------------------

---- 1 thread ----
Benchmark                   (serializationFormat)   Mode  Cnt        Score        Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  5705735,307 ± 79033,789  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  5294292,861 ± 64337,012  ops/s

---- 40 threads ----
Benchmark                   (serializationFormat)   Mode  Cnt         Score         Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  29367930,350 ± 1055106,646  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  31554336,401 ±  717383,657  ops/s

-------------------------- SpecificRecord --------------------------

---- 1 thread ----
Benchmark                   (serializationFormat)   Mode  Cnt        Score        Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  5443030,279 ± 66768,403  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  5272639,170 ± 77221,371  ops/s

---- 40 threads ----
Benchmark                   (serializationFormat)   Mode  Cnt         Score         Error  Units
SerdeBenchmark.deserialize                   AVRO  thrpt   10  29277511,317 ± 1246546,316  ops/s
SerdeBenchmark.serialize                     AVRO  thrpt   10  30689053,905 ±  743639,006  ops/s
```

Deserialization has degraded slightly compared to my incorrect caching implementation, but I would say it is negligible. In summary, when compared to the baseline performance is:

- an order of magnitude improvement for both SpecificRecord serialization and deserialization
- 1,5-2x better for GenericRecord deserialization
- GenericRecord serialization not worse off
